external-crates/move/crates/move-package/src/resolution/dependency_graph.rs

```diff
@@ -1249,15 +1257,26 @@ impl DependencyGraph {
                 .and_then(|v| v.as_array_of_tables().cloned());
         }
 
-        use std::io::Seek;
-        let mut lock = LockFile::new(
-            install_dir,
-            self.manifest_digest.clone(),
-            self.deps_digest.clone(),
-        )?;
-        lock.flush()?;
-        lock.rewind()?;
-
+        let mut lock;
+        match lock_path {
+            // Get a handle to update an existing Move.lock.
+            // Since dependency graph updates are compatible across
+            // all Move.lock schema versions, we can rely on the existing version.
+            Some(lock_path) if lock_path.exists() => {
+                lock = LockFile::from(install_dir, &lock_path)?
+            }
+            // Initialize a lock file if no existing lock_path is set for this operation.
+            _ => {
+                use std::io::Seek;
+                lock = LockFile::new(
+                    install_dir,
+                    self.manifest_digest.clone(),
+                    self.deps_digest.clone(),
+                )?;
+                lock.flush()?;
+                lock.rewind()?;
+            }
+        }
```

Comment:
Main change to stateful updates of `Move.lock` in `write_to_lock`: Use `LockFile::from` for an existing lock file, and `LockFile::new` to initialize fresh. Below this code, we perform `update_dependency_graph` in place for either case.